File: Data Structures & Algorithms/validate-parentheses/submission-9.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False
        pair = {"}":"{", "]":"[", ")":"("}
        stack = []
        for item in s:
            if item not in pair:
                stack.append(item)
            else:
                if len(stack) == 0 or pair[item] != stack[-1]:
                    return False
                stack.pop(-1)
        return len(stack) == 0

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-9.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int, int>> s;
        int max_area = 0;
        heights.push_back(0);
        int n = (int)heights.size();
        for (int i = 0; i < n; i++) {
            int extendable = i;
            while (!s.empty() and s.top().second >= heights[i]) {
                max_area = max(max_area, s.top().second * (i - s.top().first));
                extendable = s.top().first;
                s.pop();
            }
            s.push({extendable, heights[i]});
            // max_area = max(max_area, s.top().second * (n - s.top().first));
        }
        // while (!s.empty()) {
        //     max_area = max(max_area, s.top().second * (n - s.top().first));
        //     s.pop();
        // }
        return max_area;
    }
};
